(don't merge) splitted it in 2 prs

MalloZup opened the pull request:
- [x] do pr #255  

- [ ]  Create other pr after #255 is merged for other changes

MalloZup commented:
@dmacvicar @monstermunchkin  you can have a look for reviewing once you have free time. Thx in advance happy weekend

dmacvicar commented:
Can you explain some context? I have no idea of this "feature" of libvirt. What is the usecase?, motivation? implementation overview? advantages? disadvantages?

MalloZup commented:
@dmacvicar sure, i'm fixing with this pr the issue #119.

So this is more a bugfix then a new feature.

The issue was, that 

Despite "running = false" domain always running.

In the doc, we are saying: (https://github.com/dmacvicar/terraform-provider-libvirt/blob/master/website/docs/r/domain.html.markdown)running

 - (Optional) Use false to turn off the instance. If not specified, true is assumed and the instance, if stopped, will be started at next apply.

As by the issue stated, setting false has no effect. So i'm fixing this, when set to false, the domain is shut off

monstermunchkin commented on libvirt/resource_libvirt_domain.go:
```diff
@@ -28,6 +28,12 @@ type DomainMeta struct {
+// we want to shutdown the domain after creation.
+// we need to cache this var, since the running var is used for other operations.
+// if this var is true, we will destroy in the create. (this is init in read)
+var destroyDomainRequestedByUser bool
```
I would shorten this:
```
destroyDomainRequestedByUser tracks the desired state of the domain. If set to true, the domain will be destroyed after its creation.
```

MalloZup replied:
good

monstermunchkin commented on libvirt/resource_libvirt_domain_test.go:
```diff
@@ -115,6 +115,38 @@ func TestAccLibvirtDomain_Volume(t *testing.T) {
+
+			            resource "libvirt_domain" "acceptance-test-domain" {
+			                    name = "terraform-test"
+													running = false
```
Too much indentation.

monstermunchkin commented on libvirt/resource_libvirt_domain_test.go:
```diff
@@ -490,6 +522,58 @@ func testAccCheckLibvirtDomainDestroy(s *terraform.State) error {
+			return err
+		}
+		if domainRuns {
+			return fmt.Errorf("Domain should be destroyed! TEST FAILED")
```
`Domain should not be running` No need for `TEST FAILED` since `go test` will show that the test has failed.

monstermunchkin commented on libvirt/resource_libvirt_domain_test.go:
```diff
@@ -490,6 +522,58 @@ func testAccCheckLibvirtDomainDestroy(s *terraform.State) error {
+		}
+
+		if state != "shutoff" {
+			return fmt.Errorf("Domain state should be == shutoff, but is %s, TEST FAILED", state)
```
`Domain state should be 'shutoff' but is '%s'`. Again, no need for `TEST FAILED`.

dmacvicar commented:
Ok now I understand. I was confusing the resource destroy with domain destroy. The jargon for libvirt resource destroy is undefine, so it looks good to me that the domain is "destroyed" if you don't want it running.

Now, I did not implement this "running" functionality (IIRC at the beginning domains were not defined), so you may want someone that knows that concept to give it an OK.

MalloZup commented on libvirt/resource_libvirt_domain.go:
```diff
@@ -698,6 +704,8 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
+	defer destroyDomainByUserRequest(d, domain)
```
@monstermunchkin  why we don't need `defer` here? imho is more elegant to use it. 
I mean this destroying function is a `cleanup` function

dmacvicar replied:
What @monstermunchkin means it that you want to call destroyDomainByUserRequest _there_ and not later. Any programmer reading the code would think "why is he deferring something he wants to do now?".
While in the other example `defer domain.Free()` you are indeed intending "Call `domain.Free()` once all the rest of the code is executed, and _not_ before, because I will be using `domain`".

MalloZup replied:
@dmacvicar @monstermunchkin  thx for clarification. 
I put `defer` there, because we are shutting down a libvirt domain, and imho it should be added later like `defer domain.Free`, for one reason only: 
- imagine the domain is shut off (down). and people are adding after that function other operations on domain. 
The question is , are all domain operation possible on a shutdown domain without errors? 
So that's why i put this in a `defer` because to me seemed a function that need to go @ end

MalloZup commented:
@monstermunchkin thx for your review ! :sunflower: 

i was thinking that the defer function is much more elegant there.

MalloZup commented:
@monstermunchkin @dmacvicar ok all comments adressed. :dog2: thx

dmacvicar commented:
@MalloZup The fact that your function is called destroy it does not mean it is cleanup of a resource. You have no resource that you created there.

Now changing topics. I smell something fishy about `var destroyDomainRequestedByUser bool`being global. Why did you make it global?
Isn't the value going to be overwritten by two domains in the same .tf file?

I think there is a different place to store state for the resource. I'd look either in the properties itself or something around DomainMeta.

MalloZup commented:
@dmacvicar  ok thx for the catch.  

The global var is a code smell that i will address this asap.

dmacvicar commented:
More than a code smell, is a bug. It would make each domain overwrite data from the other. It is not just a question of style.

MalloZup commented:
@dmacvicar, i tried to  say it with lightness. Of course i know it that it is a bug ... that is also why i want create some tests

inercia commented on libvirt/resource_libvirt_domain.go:
```diff
@@ -28,6 +28,9 @@ type DomainMeta struct {
 var PoolSync = NewLVirtPoolSync()
 
+// destroyDomainRequestedByUser tracks the desired state of the domain. If set to true, the domain will be destroyed after its creation.
+var destroyDomainRequestedByUser bool
```
Do we really need a global here? Isn't it a domain attribute?

MalloZup replied:
@inercia  yop this i will fix, is a mistake from my part.

inercia commented on libvirt/resource_libvirt_domain.go:
```diff
@@ -1215,6 +1220,38 @@ func newDiskForCloudInit(virConn *libvirt.Connect, volumeKey string) (libvirtxml
 
+func destroyDomainByUserRequest(d *schema.ResourceData, domain *libvirt.Domain) error {
+	// if the running is true in the tf file, we don't destroy the domain
+	if !destroyDomainRequestedByUser {
```
As I said, couldn't we get the `destroyDomain` as an attribute? like `d.Get('destroyDomain')` or something like that?

MalloZup replied:
i'm thinking now how to solve this, i will address this asap

MalloZup commented:
@inercia thx for review. i will ping you guys when i am done. I'm writing a new ACC-tests and will continue it later the fix.
:crossed_fingers:

MalloZup commented on libvirt/resource_libvirt_domain_test.go:
```diff
@@ -116,6 +116,71 @@ func TestAccLibvirtDomain_Volume(t *testing.T) {
+}
+
+// we have 4 running=true and 4 running=false domain
+func TestRunningFalseDomainConcurenncyMultiDomainsRunning(t *testing.T) {
```
open for name improvements :tangerine:

MalloZup commented on libvirt/resource_libvirt_domain_test.go:
```diff
@@ -523,6 +593,103 @@ func testAccCheckLibvirtDomainDestroy(s *terraform.State) error {
 	return nil
 }
 
+func testAccCheckLibvirtDomainIsRunning(n string, domain *libvirt.Domain) resource.TestCheckFunc {
```
this  Helper Function can be refactored with the other function.

I am in favor with the generic refactoring

```golang
func testAccCheckLibvirtDomaiStatusEqual (n string, domain *libvirt.Domain, domainStatusExpected string) 
```

MalloZup commented:
so i fixed all issue.
what is new here:

- added multidomain tests for avoid concurency race condition. ( 3 domain run 3 are shut down expected)
- added test function DomainStateEqual
- the destroy is controlled over a `domain_shutoff` property and not over global var
- added documentation 

thx for reviews in advance

MalloZup commented:
@dmacvicar @inercia @monstermunchkin  so i fixed and adressed your comments.

Feel free to review it. Take your :watch:  and :tea: . I will :tent: and :running_man:  :rofl:

MalloZup commented on website/docs/r/domain.html.markdown:
```diff
@@ -31,8 +31,8 @@ The following arguments are supported:
   will be created with 512 MiB of memory be used.
-* `running` - (Optional) Use `false` to turn off the instance. If not specified,
-  true is assumed and the instance, if stopped, will be started at next apply.
+* `domain_shutoff` - (Optional) Use `true` to turn off the instance. If not specified,
```
i could change this with `domain_shutdown` open for name suggestions

monstermunchkin replied:
This is confusing:
* in the docs, you are replacing `running` with `domain_shutoff`
* however, the option `running` still exists

Why do we need both options?

MalloZup replied:
from my POV the problem was already there in the variable `running`.

We are using this property, for doing other operations, like refresh etc. The `running` is kind of internal property that we use in the domain for know if the domain is up or not, and then we do other stuff. 

The `shutoff_domain` var is only an option that we need to say that after creation we need to shutoff.

This is why i created the other var.
 I could create the runnig var in the meta Domain struct. if you want

monstermunchkin replied:
I just looked at the code, and cannot find a reason to keep the `running` option. Setting it in a `.tf` file, won't change anything. Am I missing something or is this option useless?

MalloZup replied:
i think the option is useless. i will remove it

MalloZup replied:
domain_shutoff or domain_shutdown ?

MalloZup commented:
@monstermunchkin  ok good catch :+1: :1st_place_medal: . btw we had the same `running` in network, and i removed it ( which caused also the cleanup/removal of the network_update function.) :sunflower: :sake:

MalloZup commented:
@monstermunchkin  thx  for feedback! appreciate it. 

So locally `bei mir` the test are ok. travis will take some time

MalloZup commented:
#255

MalloZup commented:
i am closing this, and i have done another pr #257
